posthog/models/sessions/sql.py

```diff
@@ -0,0 +1,159 @@
+    session_id VARCHAR,
+    -- part of order by so will aggregate correctly
+    team_id Int64,
+    -- ClickHouse will pick any value of distinct_id for the session
+    -- this is fine since even if the distinct_id changes during a session
+    -- it will still (or should still) map to the same person
+    distinct_id SimpleAggregateFunction(any, String),
+
+    min_first_timestamp SimpleAggregateFunction(min, DateTime64(6, 'UTC')),
+    max_last_timestamp SimpleAggregateFunction(max, DateTime64(6, 'UTC')),
+
+    urls SimpleAggregateFunction(groupUniqArrayArray, Array(String)),
+    entry_url AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    exit_url AggregateFunction(argMax, String, DateTime64(6, 'UTC')),
+
+    initial_referring_domain AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_source AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_campaign AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_medium AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_term AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_utm_content AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_gclid AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+    initial_gad_source AggregateFunction(argMin, String, DateTime64(6, 'UTC')),
+
+    -- create a map of how many times we saw each event
+    event_count_map AggregateFunction(sumMap, Tuple(Array(String), Array(Int64))),
+    -- duplicate the event count as a specific column for pageviews and autocaptures,
+    -- as these are used in some key queries and need to be fast
+    pageview_count SimpleAggregateFunction(sum, Int64),
+    autocapture_count SimpleAggregateFunction(sum, Int64),
+) ENGINE = {engine}
+"""
+
+SESSIONS_DATA_TABLE_ENGINE = lambda: AggregatingMergeTree(TABLE_BASE_NAME, replication_scheme=ReplicationScheme.SHARDED)
+
+SESSIONS_TABLE_SQL = lambda: (
+    SESSIONS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(min_first_timestamp)
+    -- order by is used by the aggregating merge tree engine to
+    -- identify candidates to merge, e.g. toDate(min_first_timestamp)
+    -- would mean we would have one row per day per session_id
+    -- if CH could completely merge to match the order by
+    -- it is also used to organise data to make queries faster
+    -- we want the fewest rows possible but also the fastest queries
+    -- since we query by date and not by time
+    -- and order by must be in order of increasing cardinality
+    -- so we order by date first, then team_id, then session_id
+    -- hopefully, this is a good balance between the two
+    ORDER BY (toDate(min_first_timestamp), team_id, session_id)
+SETTINGS index_granularity=512
+"""
+).format(
+    table_name=SESSIONS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=SESSIONS_DATA_TABLE_ENGINE(),
+)
+
+
+def source_column(column_name: str) -> str:
```

Comment:
I'm not sure about this, we definitely want to pull from the materialized versions of these properties if we can, but I'm not sure we can guarantee they exist even on EE

Reply:
so for context here

1) we know the query to populate the MV is faster if we use materialized columns
2) we don't know that the order of operations in CH is safe i.e. an insert of an event causes the materialized column to be computer before the select that reads that event into the sessions table

i guess we could write a script that hammers the local machine and check that we have the data we expect afterwards ??

Reply:
Please please make sure if you're adding any custom materialised columns that they have this comment like here: https://github.com/PostHog/posthog/blob/master/posthog/clickhouse/test/__snapshots__/test_schema.ambr#L482

This is used in this function: https://github.com/PostHog/posthog/blob/master/ee/clickhouse/materialized_columns/columns.py#L37 which is in turn used in `get_property_string_expr` to determine which column to use.

Without it, it's very hard to know when a column is materialised or not. Would much prefer depending on this over EE_AVAILABLE

Reply:
It's also how we get rid of the materialised-yet-or-still-migrating-problem:

https://github.com/PostHog/posthog/blob/master/ee/clickhouse/materialized_columns/columns.py#L109

The comment is added only on the materialised column after its materialised

Reply:
Ah ok! I found an example of us materializing a column in a migration! This makes this a lot more doable https://github.com/PostHog/posthog/commit/52a9054900b00d3bd8b9e7e869c49d5f785cb1f3

Reply:
Hmm, I think I don't want to add this complexity unless we can prove that it's necessary for performance. For now let's just get the property from the json object, it's easy enough to change in the future.

Reply:
Totally your call, no context on the problem here, just mentioning if you do go the materialised columns route, stick to the standard, would make your life easier

> but wouldn't elsewhere

If you decide to do it, just fyi it should work everywhere, since it checks the column COMMENT to know whether it's materialised or not, and does the default if not materialised, just like `get_property_string_expr`

Reply:
I realised that using `get_property_string_expr` here is actually good enough on its own. It'll use the materialized columns if they exist, and not if they don't. The only downside is if someone materializes these columns later, the view won't update to use them, but I think that's acceptable.